src/core/dynamic-user.c

```diff
@@ -352,42 +334,11 @@ static int dynamic_user_pop(DynamicUser *d, uid_t *ret_uid, int *ret_lock_fd) {
 
 static int dynamic_user_push(DynamicUser *d, uid_t uid, int lock_fd) {
         struct iovec iov = IOVEC_INIT(&uid, sizeof(uid));
-        union {
-                struct cmsghdr cmsghdr;
-                uint8_t buf[CMSG_SPACE(sizeof(int))];
-        } control = {};
-        struct msghdr mh = {
-                .msg_control = &control,
-                .msg_controllen = sizeof(control),
-                .msg_iov = &iov,
-                .msg_iovlen = 1,
-        };
-        ssize_t k;
 
         assert(d);
 
         /* Store the UID and lock_fd in the storage socket. This should be called with the socket pair lock taken. */
-
-        if (lock_fd >= 0) {
```

Comment:
This checked if `lock_fd` is non-negative, but is replaced with a call to `send_one_fd_iov`, which calls `send_one_fd_iov_sa`, which attaches `fd` unconditionally. Is this on purpose?

Reply:
Ah indeed this is a problem and I didn't notice it.

Right now `send_one_fd_iov()` only works when `fd >= 0` (it asserts that), so this would have to branch and use a bare `send()` if there's no FD involved.

But I'm thinking of updating `send_one_fd_iov()` so it needs either FD *or* data (which is similar to how `receive_one_fd_iov()`, which will return >= 0 if any of those are true.) I'll update this PR with that.

Other than this, what do you think of this PR in general?

Reply:
This is addressed now. `send_one_fd_iov()` can now send data with no FD attached. (But it needs to send at least one of those, otherwise it's an error.)

I added two more tests for this case.

There's one corner case here that's kind of odd... If you send a message with an empty string in it, it's considered a valid send. On the receive side though, it's hard to distinguish from receiving no FD and no data, and I'm currently returning -EIO. Should that maybe return 0 instead, signaling that no data was received?

PTAL.